File: src/main/java/com/matag/admin/cleanup/MatagGameCleanup.java
package com.matag.admin.cleanup;

import com.matag.admin.game.game.Game;
import com.matag.admin.game.game.GameRepository;
import com.matag.admin.game.session.GameSession;
import com.matag.admin.game.session.GameSessionRepository;
import lombok.AllArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.time.Clock;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import static com.matag.admin.game.game.GameStatusType.IN_PROGRESS;
import static com.matag.admin.game.game.GameStatusType.STARTING;

@Component
@AllArgsConstructor
public class MatagGameCleanup {
  private static final Logger LOGGER = LoggerFactory.getLogger(MatagGameCleanup.class);

  private final GameRepository gameRepository;
  private final GameSessionRepository gameSessionRepository;
  private final Clock clock;

  @Transactional
  public void cleanup() {
    cleanupOldUnfinishedGames();
    cleanupFinishedGuestGames();
  }

  private void cleanupOldUnfinishedGames() {
    var oldNotFinishedGames = gameRepository.findOldGameByStatus(List.of(STARTING, IN_PROGRESS), LocalDateTime.now(clock).minusDays(2));
    LOGGER.info("cleanupOldUnfinishedGames...");
    deleteGames(oldNotFinishedGames);
  }

  private void cleanupFinishedGuestGames() {
    LOGGER.info("cleanupFinishedGuestGames...");
    var oldGuestGamesIds = gameRepository.findFinishedGuestGames(LocalDateTime.now(clock).minusDays(2));
    var oldGuestGames = oldGuestGamesIds.stream()
      .map(gameRepository::findById)
      .flatMap(Optional::stream)
      .collect(Collectors.toList());
    deleteGames(oldGuestGames);
  }

  private void deleteGames(List<Game> games) {
    LOGGER.info("MatagGameCleanup identified " + games.size() + " entries to delete.");
    var totalGameDeleted = 0;
    for (var game : games) {
      try {
        for (GameSession gameSession : game.getGameSessions()) {
          gameSessionRepository.delete(gameSession);
          gameRepository.delete(game);
          totalGameDeleted++;
        }

      } catch (Exception e) {
        LOGGER.error("Could not delete game with id " + game.getId(), e);
      }
    }

    LOGGER.info("MatagGameCleanup deleted " + totalGameDeleted + " games");
  }
}

File: src/main/kotlin/com/matag/admin/cleanup/MatagGameCleanup.kt
package com.matag.admin.cleanup

import com.matag.admin.game.game.Game
import com.matag.admin.game.game.GameRepository
import com.matag.admin.game.game.GameStatus
import com.matag.admin.game.session.GameSessionRepository
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.time.Clock
import java.time.LocalDateTime

@Component
open class MatagGameCleanup(
  private val gameRepository: GameRepository,
  private val gameSessionRepository: GameSessionRepository,
  private val clock: Clock
) {

    @Transactional
    open fun cleanup() {
        cleanupOldUnfinishedGames()
        cleanupFinishedGuestGames()
    }

    private fun cleanupOldUnfinishedGames() {
        val oldNotFinishedGames = gameRepository.findOldGameByStatus(
            listOf(
                GameStatus.STARTING,
                GameStatus.IN_PROGRESS
            ), LocalDateTime.now(clock).minusDays(2)
        )
        LOGGER.info("cleanupOldUnfinishedGames...")
        deleteGames(oldNotFinishedGames)
    }

    private fun cleanupFinishedGuestGames() {
        LOGGER.info("cleanupFinishedGuestGames...")
        val oldGuestGamesIds = gameRepository.findFinishedGuestGames(LocalDateTime.now(clock).minusDays(2))
        val oldGuestGames = oldGuestGamesIds
            .map { gameRepository.findById(it) }
            .map { it.get() }
        deleteGames(oldGuestGames)
    }

    private fun deleteGames(games: List<Game>) {
        LOGGER.info("MatagGameCleanup identified " + games.size + " entries to delete.")
        var totalGameDeleted = 0
        for (game in games) {
            try {
                for (gameSession in game.gameSessions) {
                    gameSessionRepository.delete(gameSession)
                    gameRepository.delete(game)
                    totalGameDeleted++
                }
            } catch (e: Exception) {
                LOGGER.error("Could not delete game with id " + game.id, e)
            }
        }

        LOGGER.info("MatagGameCleanup deleted " + totalGameDeleted + " games")
    }

    companion object {
        private val LOGGER: Logger = LoggerFactory.getLogger(MatagGameCleanup::class.java)
    }
}
